### PR TITLE
GitHub xrays [WIP]

### PR DESCRIPTION
**Pull request by sbelak**
# GitHub X-rays (via Singer.io)

## Getting this to work

Use the Singer.io tap https://github.com/singer-io/tap-github and pipe it into a DB of your choice. The process will probably fail a bunch of times. I found it best to add streams one by one. It will also take quite some time (at least if you're dumping the Metabase repo). Make sure the FKs are in place, if not add them manually. After this, Metabase should identify the relevant tables as being GitHuby and apply appropriate x-rays (as of now: Issues, PRs, Commits, Stargazers)

## Example
Screenshot 2019-03-12 21 10 41
Screenshot 2019-03-12 21 12 28

## Limitations

3 big (& non-trivial) limitations jump from this:

1) Table is the worng abstraction level for xray definitions. GH xray is screaming for an overview (and in fact a single dashboard would suffice if done right), instead we have a bunch of pretty small and barren per-table xrays/dashboards.

2) x-rays should support nested questions. This would allow questions such as "PRs by first-time contributors" (and in general mitigate the limitation that you can't filter on a metric directly as well as act as pseudo-ETL for when targeting fixed schemas that are not ideal).

3) If we're going to have a lot of datasource-specific x-rays, table `:entity-type` classification should be done on the schema/DB level, rather than per-table. Also binding by `entity-type` in itself is problematic and we'll soon have name (& concept) clashes. There should probably be a way to specify a table name or pattern, same as we do with fields (`:named`).

Minor issues:
* would be nice to be able to stack cards
* count-when (which is already the baseline for share but not exposed) would be useful
* Related sometimes gets polluted with low-relevance tables (we should probably tweak the recommendations such that when there are candidates with domain `entity-type`s, don't show any generic)
* There should be more folder structure (ie. all the GH heuristics should be in a separate folder)

## (general x-ray) next steps

1) Relax the heuristic-schema correspondence
2) Rethink the role of `entity-type` and how we bind tables. 
3) Rethink heuristics folder organization with an eye towards extensibility
4) nested questions
5) nicer validation errors and more validation tools to make writing new heuristics nicer

### CODE REVIEW

**Comment by sbelak** [comment]
General impression writing these was that it's much less annoying as you have a fixed target and aren't battling combinatorial explosions. There are a couple more shortcuts that could be made and nicer validation errors, but overall it's not that bad.

**Comment by sbelak** [comment]
Next steps: #9564, #9565

File: src/metabase/types.clj [hunk 1]
```diff
@@ -22,8 +22,14 @@
 (derive :entity/ProductTable :entity/GenericTable)
 (derive :entity/SubscriptionTable :entity/GenericTable)
 (derive :entity/EventTable :entity/GenericTable)
+
 (derive :entity/GoogleAnalyticsTable :entity/GenericTable)
 
+(derive :entity/GHCommitsTable :entity/GenericTable)
```
Comment (salsakran) [inline src/metabase/types.clj]: GH is too cryptic here ... as we add data types, there will be multiple domains where "GH" could come from.

**Review by salsakran — CHANGES_REQUESTED** [review]
in general not liking the use of specific field names in the yaml file vs resolving them via the type system.

File: src/metabase/types.clj (hunk 1 above)
Comment (sbelak) [reply src/metabase/types.clj]: Agree, this is temporary to make it work with what we have. #9564 outlines a more elegant approach, where not every entity type needs to be global.

File: resources/automagic_dashboards/table/GHCommitsTable.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+title: GitHub Commits
+transient_title: "A summary of your commits"
+metrics:
+  - Count: ["count"]
+  - CumCount: ["cum-count", [dimension, Sha]]
+dimensions:
+ - CommitedAt:
+     field_type: DateTime
+     named: commit__author__date
+ - Sha:
```
[the remark this answers is not in the document]
Comment (sbelak) [reply resources/automagic_dashboards/table/GHCommitsTable.yaml]: Continuing this in #9565

**Comment by sbelak** [comment]
To consolidate the conversations we had:
* name based binding is a bad idea, we should rather invest more effort into our type system 
* There should be tighter correspondence between and awareness off field and table types
* There should be a couple of verticals our typesystem covers really well, but extensible type system is not a priority. 

Consequently 2 things need to happen:
1) type classification needs to become more robust & expressive
2) type system needs to become more hierarchal/namespacey

**Comment by sbelak** [comment]
Pushed all name-based binding to the type system. This introduces a couple of problems:

* We can no longer filter on state to `opened` for issues.
Solution 1: make geographical state classifier peak at values (this one might suck performance-wise, especially since we'd have to test for all countries with states).
Solution 2: provide another state-like type (status?) which can be manually set.
* It's no longer possible to generate links or even just issue #.
(partial) Solution: add type `InternalID` which can be manually mapped in such circumstances (Singer uses exceedingly dumb name `number` for this field so hard to see how this could be done automatically).
* There are some nonsensical duplicates as multiple fields match
Solution: add option to say "there will be at most one such field" to the binding logic. Can also partially be solved by manually removing erroneous type classifications and or hide those fields (which tend to be on cruft anyhow)
* We loose some of the stats fields like `comments` on issues
Solution 1: add stuff like "number of comments" to the type system as well
Solution 2: yolo and treat all number fields as relevant (and perhaps add some cruft detection for fields as well)

**Todo** 
- [ ] Make FE aware of the new types

**Comment by rlotun** [comment]
closing due to age.

**Pull request closed without merging**